File: Expt 2/c.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#define MAX 100

long int lstack[MAX];
float fstack[MAX];
char postfix[MAX];
int top = -1;
int floatint;

void push(float);
float pop();
void evalpostfix();
int isfull();
int isempty();

int isempty(){
	if(top == -1)
		return 1;
	else return 0;
}

int isfull(){
	if(top == (MAX - 1))
		return 1;
	else return 0;
}

void push(float symbol){
	if(isfull()){
		printf("Stack overflow\n");
		return;
	}
	if (floatint == 1)
		fstack[++top] = symbol;
	else lstack[++top] = (long int)symbol;
}

float pop(){
	if(isempty()){
		printf("Stack underflow\n");
		exit(1);
	}
	if (floatint == 1)
		return(fstack[top--]);
	else return((long int)lstack[top--]);
}

void evalpostfix(){
	float a, b;
	float temp;
	for(int i = 0; i < strlen(postfix); i++){
		if(postfix[i] >= '0' && postfix[i] <= '9')
			push((float)(postfix[i] - '0'));
		else{
			b = pop();
			if (floatint == 1) {
				printf("%f\t", b);
   				for (int j = 0; j <= top; j++)
        		printf("%f,", fstack[j]);
			} else {
				printf("%ld\t", b);
    			for (int j = 0; j <= top; j++)
        			printf("%ld,", lstack[j]);
			}
			printf("\n");
			a = pop();
			if (floatint == 1) {
				printf("%f\t", a);
   				for (int j = 0; j <= top; j++)
        		printf("%f,", fstack[j]);
			} else {
				printf("%ld\t", a);
    			for (int j = 0; j <= top; j++)
        			printf("%ld,", lstack[j]);
			}
			printf("\n");
			switch(postfix[i]){
				case '+':
					temp = b + a;
					break;
				case '-':
					temp = b - a;
					break;
				case '*':
					temp = b * a;
					break;
				case '/':
					temp = b / a;
					break;
				case '%':
					temp = (long int)b % (long int)a;
					break;
				case '^':
					temp = 1;
					for(int i = 1; i <= a; i++)
						temp = temp * b;
					break;
				case 'A': 
					temp = 3;
					break;
				case 'B':
				case 'C':
				case 'D':
				case 'E':
					temp = 2;
					break;
				case 'F':
					temp = 1;
					break;
			}
			push(temp);
		}
	}
}

int main(){
	printf("Enter the postfix expression: ");
	scanf("%s", postfix);
	do{
		printf("\nDo you want it as a 1. Float or 2. Integer?: ");
		scanf("%d", &floatint);
		if (floatint != 1 && floatint != 2)
			printf("Invalid input\n");
	} while (floatint != 1 && floatint != 2);
	printf("Symbol\tStack\t\n");
	evalpostfix();
	if(floatint == 1)
		printf("The evaluation is: %f\n", fstack[0]);
	else 
		printf("The evaluation is: %f\n", lstack[0]);
	return 0;
}

File: Expt 2/c.py
def precedence(operator):
    if operator == '^':
        return 3
    elif operator in ['*', '/']:
        return 2
    elif operator in ['+', '-']:
        return 1
    else:
        return 0

def infix_to_postfix(expression):
    postfix = []
    stack = []

    for symbol in expression:
        if symbol.isalnum():
            postfix.append(symbol)
        elif symbol == '(':
            stack.append(symbol)
        elif symbol == ')':
            while stack and stack[-1] != '(':
                postfix.append(stack.pop())
            stack.pop()  # Remove the '('
        else:
            while stack and precedence(stack[-1]) >= precedence(symbol):
                postfix.append(stack.pop())
            stack.append(symbol)

    while stack:
        postfix.append(stack.pop())

    return ''.join(postfix)

def evaluate_postfix(expression, values, output_type):
    stack = []

    for symbol in expression:
        if symbol.isalnum():
            stack.append(values[symbol])
        else:
            operand2 = stack.pop()
            operand1 = stack.pop()
            if symbol == '+':
                stack.append(operand1 + operand2)
            elif symbol == '-':
                stack.append(operand1 - operand2)
            elif symbol == '*':
                stack.append(operand1 * operand2)
            elif symbol == '/':
                stack.append(operand1 / operand2)
            elif symbol == '^':
                stack.append(operand1 ** operand2)

    if output_type == float:
        stack = [round(item, 6) if isinstance(item, float) else item for item in stack]

    return stack[0]

infix_expression = "(F-D/(C+A)*B*B-D^E^A)"
values = {'A': 3, 'B': 2, 'C': 2, 'D': 2, 'E': 2, 'F': 1}

# Convert infix expression to postfix
postfix_expression = infix_to_postfix(infix_expression)
print("Symbol\tStack\tPostfix")
print("----------------------------")
stack = []
for symbol in infix_expression:
    if symbol == '(':
        stack.append(symbol)
    elif symbol == ')':
        while stack and stack[-1] != '(':
            print(f"{symbol}\t{' '.join(map(str, stack))}\t{postfix_expression}")
            stack.pop()
        stack.pop()
    elif symbol.isalnum():
        print(f"{symbol}\t{' '.join(map(str, stack))}\t{postfix_expression}")
    else:
        while stack and precedence(stack[-1]) >= precedence(symbol) and stack[-1] != '(':
            print(f"{symbol}\t{' '.join(map(str, stack))}\t{postfix_expression}")
            stack.pop()
        stack.append(symbol)

while stack:
    print(f"\t{' '.join(map(str, stack))}\t{postfix_expression}")
    stack.pop()

print("\nPostfix Expression:", postfix_expression)

print("Enter the postfix expression: ")
print(infix_expression)
# User option for result type
output_option = input("\nDo you want the result as a: 1. Float or 2. Integer?: 1")

output_type = float if output_option == '1' else int

# Evaluate postfix expression
result = evaluate_postfix(postfix_expression, values, output_type)
print("\nSymbol\tStack")
stack = []
for symbol in postfix_expression:
    if symbol.isalnum():
        stack.append(values[symbol])
    else:
        operand2 = stack.pop()
        operand1 = stack.pop()
        if symbol == '+':
            result_str = operand1 + operand2
        elif symbol == '-':
            result_str = operand1 - operand2
        elif symbol == '*':
            result_str = operand1 * operand2
        elif symbol == '/':
            result_str = operand1 / operand2
        elif symbol == '^':
            result_str = operand1 ** operand2

        stack.append(result_str)
    stack_display = ['%.6f' % item if isinstance(item, float) else item for item in stack]
    print(f"{symbol}\t{' '.join(map(str, stack_display))}")

print("\nThe evaluation is:", result)
